File: java/src/main/java/money_problem/domain/ConversionResult.java
package money_problem.domain;

public record ConversionResult<Failure>(Money money, Failure failure) {
    private ConversionResult(Money money) {
        this(money, null);
    }

    private ConversionResult(Failure exception) {
        this(null, exception);
    }

    public static <F> ConversionResult<F> fromFailure(F failure) {
        return new ConversionResult<>(failure);
    }

    public static <F> ConversionResult<F> fromSuccess(Money money) {
        return new ConversionResult<>(money);
    }

    public boolean isFailure() {
        return failure != null;
    }

    public boolean isSuccess() {
        return money != null;
    }
}

File: c#/money-problem.Domain/ConversionResult.cs
namespace money_problem.Domain;

public class ConversionResult<T>
{
    public ConversionResult(Money money)
    {
        this.Money = money;
    }

    public ConversionResult(T failure)
    {
        this.Failure = failure;
    }

    public Money? Money { get; }

    public T? Failure { get; }

    public bool IsFailure() => this.Failure is { };

    public bool IsSuccess() => this.Money is { };

    public static ConversionResult<T> FromFailure(T failure) => new ConversionResult<T>(failure);

    public static ConversionResult<T> FromMoney(Money money) => new ConversionResult<T>(money);
}